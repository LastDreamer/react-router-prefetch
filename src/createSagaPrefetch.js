export default (props, type, payload) => new Promise(
  (resolve, reject) => props.dispatch({
    type,
    payload,
    resolve,
    reject,
  }),
);
