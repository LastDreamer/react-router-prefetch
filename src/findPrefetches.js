import { Children } from 'react';

const getNodeType = (node) => {
  if (typeof node.type === 'function') {
    if (node.type.prototype.isReactComponent ||
        node.type.prototype.isPureReactComponent) {
      return 'statefull';
    }
    return 'stateless';
  }

  if (!node.type && node) {
    return node.constructor.name.toLocaleLowerCase();
  }

  return node.type.constructor.name.toLocaleLowerCase();
};

const initComponent = (node, context) => {
  switch (getNodeType(node)) {
    case 'statefull': {
      // eslint-disable-next-line new-cap
      const element = new node.type(node.props, context);

      if (element.componentWillMount) {
        try {
          element.componentWillMount();
        } catch (e) {
          // eslint-disable-next-line no-console
          console.log('error');
        }
      }

      let component;

      try {
        component = element.render();
      } catch (e) {
        component = [];
      }

      return [
        component,
        Object.assign(
          context,
          element.getChildContext
            ? element.getChildContext()
            : {}),
      ];
    }
    case 'stateless':
      return [
        node.type(node.props, context),
        context,
      ];
    default:
      return [
        node,
        context,
      ];
  }
};

const findPrefetches = (node, context = {}, prefetchMethod = 'prefetch') => {
  let prefetches = [];

  if (!node) {
    return prefetches;
  }

  if (node.type &&
      node.type[prefetchMethod] &&
      !node.type.WrappedComponent &&
      !node.type.Naked
  ) {
    prefetches.push([node.type[prefetchMethod], node.props]);
  }

  const [component, childContext] = initComponent(node, context);

  if (component) {
    const nodeType = getNodeType(component);

    if (nodeType === 'stateless' || nodeType === 'statefull') {
      prefetches = prefetches.concat(
        findPrefetches(component, context, prefetchMethod),
      );
    } else {
      let children;

      if (nodeType === 'array') {
        children = component;
      } else if (component.props && component.props.children) {
        // eslint-disable-next-line prefer-destructuring
        children = component.props.children;
      }

      if (children) {
        Children.forEach(children, (child) => {
          const childPrefetches = findPrefetches(
            child,
            childContext,
            prefetchMethod,
          );

          if (childPrefetches.length) {
            prefetches = prefetches.concat(childPrefetches);
          }
        });
      }
    }
  }

  return prefetches;
};

export default findPrefetches;
