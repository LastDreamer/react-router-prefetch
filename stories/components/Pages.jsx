import React, { Fragment } from 'react';
import { Route } from 'react-router-dom';

import Page1 from './Page1';
import Page2 from './Page2';
import Page3 from './Page3';
import Page4 from './Page4';
import Page5 from './Page5';

export default () => (
  <Fragment>
    <Route path="/one" component={Page1} />
    <Route path="/two" component={Page2} />
    <Route path="/three" component={Page3} />
    <Route path="/four" component={Page4} />
    <Route path="/five" component={Page5} />
  </Fragment>
);
