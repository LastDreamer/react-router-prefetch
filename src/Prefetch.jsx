import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withRouter, StaticRouter } from 'react-router-dom';
import defaultsDeep from 'lodash/defaultsDeep';

import findPrefetches from './findPrefetches';

class Prefetch extends Component {
  static contextTypes = {
    store: PropTypes.any,
    router: PropTypes.object.isRequired,
  }

  static childContextTypes = {
    store: PropTypes.any,
    router: PropTypes.object.isRequired,
  }

  state = {
    initialHide: this.props.initialHide,
  }

  getChildContext = () => defaultsDeep({
    router: {
      route: {
        location: this.state.location,
      },
    },
  }, this.context)

  componentDidMount = () => this.checkPrefetches(this.props)

  componentWillReceiveProps = this.checkPrefetches

  transitionsCount = 0

  checkPrefetches({ children, location, prefetchMethod }) {
    const { createHref } = this.context.router.history;

    if (this.state.location &&
        createHref(location) === createHref(this.state.location)) {
      return;
    }

    this.transitionsCount += 1;

    const context = defaultsDeep({
      router: {
        location,
      },
    }, this.getChildContext());

    const prefetches = findPrefetches(
      <StaticRouter
        context={this.getChildContext()}
        location={location}
      >
        {React.cloneElement(children)}
      </StaticRouter>,
      context,
      prefetchMethod,
    );

    if (prefetches.length) {
      this.setState({
        prefetches,
        fetchRequested: true,
        nextLocation: location,
      }, this.prefetchRoutes);
    } else {
      this.setState({
        initialHide: false,
        fetchRequested: false,
        location,
      });
    }
  }

  prefetchRoutes() {
    const {
      onFetchStart,
      onFetchEnd,
      errorMessage,
      onError,
    } = this.props;

    onFetchStart();

    const currentTransition = this.transitionsCount;

    Promise
      .all(this.state.prefetches.map(
        ([getPromise, props]) => getPromise(props),
      ))
      .then(() => {
        if (currentTransition === this.transitionsCount) {
          this.setState({
            initialHide: false,
            prefetches: [],
            fetchRequested: false,
            location: this.state.nextLocation,
          }, () => {
            onFetchEnd();
          });
        }
      })
      .catch((error) => {
        if (currentTransition === this.transitionsCount) {
          this.setState({
            fetchRequested: false,
            prefetches: [],
            nextLocation: this.state.location,
          }, () => {
            onError(errorMessage, error);
            onFetchEnd();
          });
        }
      });
  }

  render() {
    const { children, preloader } = this.props;
    const { fetchRequested, initialHide } = this.state;

    return (
      <Fragment>
        { fetchRequested && preloader }
        { !(fetchRequested && initialHide) && children }
      </Fragment>
    );
  }
}

Prefetch.propTypes = {
  children: PropTypes.node.isRequired,
  errorMessage: PropTypes.string,
  initialHide: PropTypes.bool,
  onError: PropTypes.func.isRequired,
  onFetchStart: PropTypes.func,
  onFetchEnd: PropTypes.func,
  // eslint-disable-next-line react/no-unused-prop-types
  prefetchMethod: PropTypes.string,
  preloader: PropTypes.node,
};

Prefetch.defaultProps = {
  errorMessage: 'Error while page loading',
  initialHide: true,
  onFetchStart: () => {},
  onFetchEnd: () => {},
  prefetchMethod: 'prefetch',
  preloader: 'Loading...',
};

export const PrefetchComponent = Prefetch;

export default withRouter(Prefetch);
