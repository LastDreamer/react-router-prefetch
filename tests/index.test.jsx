/* global document */
import React, { Component } from 'react';
import ReactDom from 'react-dom';
import { MemoryRouter, Router } from 'react-router';
import ShallowRenderer from 'react-test-renderer/shallow';
import { mount, shallow } from 'enzyme';
import { createMemoryHistory } from 'history';

import Pages from '../stories/components/Pages';
import findPrefetches from '../src/findPrefetches';
import createSagaPrefetch from '../src/createSagaPrefetch';
import Prefetch from '../src';

describe('Prefetch', () => {
  it('should render itself', () => {
    const wrapper = shallow(
      <Prefetch
        onError={() => {}}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should show children if prefetch not needed', () => {
    const wrapper = mount(
      <MemoryRouter
        initialEntries={['/two']}
      >
        <Prefetch
          onError={() => {}}
        >
          <Pages />
        </Prefetch>
      </MemoryRouter>,
    );

    expect(wrapper.find('h1').length).toBe(1);
  });

  it('should not show children if prefetch needed', () => {
    const wrapper = mount(
      <MemoryRouter
        initialEntries={['/one']}
      >
        <Prefetch
          onError={() => {}}
        >
          <Pages />
        </Prefetch>
      </MemoryRouter>,
    );

    expect(wrapper.find('h1').length).toBe(0);
  });

  it('should show children if prefetch needed and initialHide are false', () => {
    const wrapper = mount(
      <MemoryRouter
        initialEntries={['/one']}
      >
        <Prefetch
          onError={() => {}}
          initialHide={false}
        >
          <Pages />
        </Prefetch>
      </MemoryRouter>,
    );

    expect(wrapper.find('h1').length).toBe(1);
  });

  it('should call onFetchStart callback', () => {
    const node = document.createElement('div');
    const spy = jest.fn();
    ReactDom.render(
      <MemoryRouter
        initialEntries={['/one']}
      >
        <Prefetch
          onError={() => {}}
          onFetchStart={spy}
        >
          <Pages />
        </Prefetch>
      </MemoryRouter>,
      node,
    );

    expect(spy).toHaveBeenCalled();
  });

  it('should call onFetchEnd callback', () => {
    const node = document.createElement('div');
    const spy = jest.fn();
    ReactDom.render(
      <MemoryRouter
        initialEntries={['/three']}
      >
        <Prefetch
          onError={() => {}}
          onFetchEnd={spy}
        >
          <Pages />
        </Prefetch>
      </MemoryRouter>,
      node,
    );

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(spy).toHaveBeenCalled();
        resolve();
      }, 1001);
    });
  });

  it('should call onError callback', () => {
    const node = document.createElement('div');
    const spy = jest.fn();

    ReactDom.render(
      <MemoryRouter
        initialEntries={['/four']}
      >
        <Prefetch
          onError={spy}
        >
          <Pages />
        </Prefetch>
      </MemoryRouter>,
      node,
    );

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(spy).toHaveBeenCalled();
        resolve();
      }, 1501);
    });
  });

  it('should not call callbacks of got same path', () => {
    const node = document.createElement('div');
    const onStart = jest.fn();
    const onEnd = jest.fn();
    const history = createMemoryHistory();

    ReactDom.render(
      <Router
        initialEntries={['/three']}
        history={history}
      >
        <Prefetch
          onError={() => {}}
          onFetchStart={onStart}
          onFetchEnd={onEnd}
        >
          <Pages />
        </Prefetch>
      </Router>,
      node,
    );

    history.push({ pathname: '/three' });

    return new Promise((resolve) => {
      setTimeout(() => {
        expect(onStart).toHaveBeenCalledTimes(1);
        expect(onEnd).toHaveBeenCalledTimes(1);
        resolve();
      }, 2001);
    });
  });
});


describe('findPrefetches', () => {
  it('should get prefetches if exists', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <MemoryRouter
        initialEntries={['/three']}
      >
        <Pages />
      </MemoryRouter>,
    );
    expect(findPrefetches(renderer.getRenderOutput()).length).toBe(1);
  });

  it('should not get prefetches if not exists', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <MemoryRouter
        initialEntries={['/two']}
      >
        <Pages />
      </MemoryRouter>,
    );
    expect(findPrefetches(renderer.getRenderOutput()).length).toBe(0);
  });

  it('should get prefetches recursive', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <MemoryRouter
        initialEntries={['/one']}
      >
        <Pages />
      </MemoryRouter>,
    );
    expect(findPrefetches(renderer.getRenderOutput()).length).toBe(2);
  });

  it('should get methods with custom names', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <MemoryRouter
        initialEntries={['/four']}
      >
        <Pages />
      </MemoryRouter>,
    );

    expect(findPrefetches(
      renderer.getRenderOutput(),
      {},
      'customPrefetch',
    ).length).toBe(1);
  });

  it('should not crash the app if nothing was passed', () => {
    expect(findPrefetches().length).toBe(0);
  });

  it('should not app when componentWillMount returns error', () => {
    class ErrorComponent extends Component {
      componentWillMount() {
        throw new Error('hello');
      }

      render() {
        return (
          <div />
        );
      }
    }
    const renderer = new ShallowRenderer();
    renderer.render(
      <MemoryRouter
        initialEntries={['/four']}
      >
        <Pages />
        <ErrorComponent />
      </MemoryRouter>,
    );

    expect(() => {
      findPrefetches(renderer.getRenderOutput());
    }).not.toThrow();
  });
});

describe('createSagaPrefetch', () => {
  it('should return Promise', () => {
    const prefetch = createSagaPrefetch(
      {
        dispatch: () => {},
      },
      'type',
      () => {},
      () => {},
    );

    expect(prefetch.constructor.name).toBe('Promise');
  });
});
