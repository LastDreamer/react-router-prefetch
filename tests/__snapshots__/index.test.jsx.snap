// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prefetch should render itself 1`] = `
<Route
  render={[Function]}
/>
`;
